sql/core/src/test/scala/org/apache/spark/sql/CollationExpressionWalkerSuite.scala

```diff
@@ -0,0 +1,294 @@
+      val headConstructor = cl.getConstructors.head
+
+      val params = headConstructor.getParameters.map(p => p.getType)
+
+      val args = generateData(params.toSeq, Utf8Binary)
+      // Find all expressions that have string as input
+      try {
+        val expr = headConstructor.newInstance(args: _*)
+        expr match {
+          case types: ExpectsInputTypes =>
+            expectsExpressionCounter = expectsExpressionCounter + 1
+            val inputTypes = types.inputTypes
+            inputTypes.exists(hasStringType) || inputTypes.isEmpty
+        }
+      } catch {
+        case _: Throwable => false
+      }
+    }).toArray
+
+    val toSkip = List(
+      "parse_url", // Parse URL cannot be generalized with ExpectInputTypes
+      "hex" // Different inputs affect conversion
+    )
+    // scalastyle:off println
+    println("Total number of expression: " + expressionCounter)
+    println("Total number of expression that expect input: " + expectsExpressionCounter)
+    println("Number of extracted expressions of relevance: " + (funInfos.length - toSkip.length))
+    // scalastyle:on println
+    for (f <- funInfos.filter(f => !toSkip.contains(f.getName))) {
+      val cl = Utils.classForName(f.getClassName)
+      val headConstructor = cl.getConstructors.head
+      val params = headConstructor.getParameters.map(p => p.getType)
+      val args = generateData(params.toSeq, Utf8Binary)
+      val expr = headConstructor.newInstance(args: _*).asInstanceOf[ExpectsInputTypes]
+      val inputTypes = expr.inputTypes
+
+      val inputDataUtf8Binary =
+        generateData(replaceExpressions(inputTypes, params.toSeq), Utf8Binary)
+      val instanceUtf8Binary =
+        headConstructor.newInstance(inputDataUtf8Binary: _*).asInstanceOf[Expression]
+
+      val inputDataLcase =
+        generateData(replaceExpressions(inputTypes, params.toSeq), Utf8BinaryLcase)
+      val instanceLcase = headConstructor.newInstance(inputDataLcase: _*).asInstanceOf[Expression]
+
+      val exceptionUtfBinary = {
+        try {
+          instanceUtf8Binary.eval(EmptyRow)
+          None
+        } catch {
+          case e: Throwable => Some(e)
+        }
+      }
+
+      val exceptionLcase = {
+        try {
+          instanceLcase.eval(EmptyRow)
+          None
+        } catch {
+          case e: Throwable => Some(e)
+        }
+      }
+
+      // Check that both cases either throw or pass
+      assert(exceptionUtfBinary.isDefined == exceptionLcase.isDefined)
+
+      if (exceptionUtfBinary.isEmpty) {
+        val resUtf8Binary = instanceUtf8Binary.eval(EmptyRow)
+        val resUtf8Lcase = instanceLcase.eval(EmptyRow)
+
+        val dt = instanceLcase.dataType
+
+        dt match {
+          case _: StringType if resUtf8Lcase != null && resUtf8Lcase != null =>
+            assert(resUtf8Binary.isInstanceOf[UTF8String])
+            assert(resUtf8Lcase.isInstanceOf[UTF8String])
+            // scalastyle:off caselocale
```

Comment:
although you're only dealing with ASCII strings here, just a heads up: we should probably be careful regarding which toLowerCase we use here (i.e. JVM vs ICU)

since UTF8_BINARY_LCASE mostly works on code point level (since https://github.com/apache/spark/pull/46700), if you do ever use non-ASCII strings in this test - consider using `CollationAwareUTF8String.lowerCaseCodePoints` instead of `UTF8String. toLowerCase`